jobs: TestRetriesWithExponentialBackoff is flaky under stress

This was on a 23.2 backport.

https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelExtendedCi/12273030?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true

```
No pass/skip/fail event found for test
=== RUN   TestRetriesWithExponentialBackoff/pause_running
    registry_test.go:822: condition failed to evaluate within 45s: from registry_test.go:830: waiting job 910125136442097665 to reach a terminal state, currently running
panic: send on closed channel [recovered]
  panic: send on closed channel
goroutine 34997 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xbe?, {0x6e34360, 0xc00c22b170})
  github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x50db2e0, 0x6dec650})
  GOROOT/src/runtime/panic.go:884 +0x213
github.com/cockroachdb/cockroach/pkg/jobs.TestRetriesWithExponentialBackoff.func6.2.1({0x6e34360, 0xc00be33dd0})
  github.com/cockroachdb/cockroach/pkg/jobs/registry_test.go:737 +0xb5
github.com/cockroachdb/cockroach/pkg/jobs/jobstest.FakeResumer.Resume({0xc00899c2e0, 0xc00899c2f0, 0x0, 0x0, 0x0}, {0x6e34360?, 0xc00be33dd0?}, {0xc0093c4360?, 0x0?})
  github.com/cockroachdb/cockroach/pkg/jobs/jobstest/resumer.go:39 +0x50
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2(0xc00dfc2b60?, 0x0?, 0xc004ccb810, {0x6e32130?, 0xc00c22b110?}, {0x6e34360?, 0xc00be33dd0?}, {0x563fd60?, 0xc004c35eb8?})
  github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1664 +0xcc
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc004863900, {0x6e34360, 0xc00c831440}, {0x563fd60, 0xc004c35eb8}, {0x6e32130, 0xc00c22b110}, 0xc00dfc2b60, {0xc00c61f8b8, 0x7}, ...)
  github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1665 +0xa3f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc004863900, {0x6e34360, 0xc00c831440}, {0x6e32130, 0xc00c22b110}, 0xc00dfc2b60, {0xc00c61f8b8, 0x7}, {0xc0091e5488, 0x16})
  github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:477 +0x556
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func2({0x6e34360?, 0xc00c22b170?})
  github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:389 +0x146
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
  github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x146
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
  github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x43b
```

When I stress it locally I get a 3rd kind of failure:

```
--- FAIL: TestRetriesWithExponentialBackoff (52.72s)
    test_log_scope.go:170: test logs captured to: /tmp/_tmp/36d09785a9c246f79b738d896ab18d71/logTestRetriesWithExponentialBackoff3965307374
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:111: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    conditional_wrap.go:190: 
        pkg/jobs/registry_test.go:720: (TestRetriesWithExponentialBackoff.func6)
                NOTICE: .InternalDB() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().InternalDB() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    test_server_shim.go:111: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    conditional_wrap.go:190: 
        pkg/jobs/registry_test.go:720: (TestRetriesWithExponentialBackoff.func6)
                NOTICE: .InternalDB() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().InternalDB() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    test_server_shim.go:111: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    test_server_shim.go:217: scheduled job created without a cluster ID
    --- FAIL: TestRetriesWithExponentialBackoff/revert_on_fail (0.81s)
        testing.go:1490: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
    panic.go:523: -- test log scope end --
FAIL
I231030 17:23:57.732910 1 (gostd) testmain.go:229  [-] 1  Test //pkg/jobs:jobs_test exited with error code 1
```

--- a/pkg/jobs/registry.go
+++ b/pkg/jobs/registry.go
@@ -1463,6 +1463,30 @@ func getRegisterOptions(typ jobspb.Type) (registerOptions, bool) {
 	return opts, ok
 }
 
+// TestingClearConstructors clears all previously registered
+// constructors. This is useful in tests when you want to ensure that
+// the job system will only run a particular job.
+//
+// The returned function should be called at the end of the test to
+// restore the constructors.
+func TestingClearConstructors() func() {
+	globalMu.Lock()
+	defer globalMu.Unlock()
+
+	oldConstructors := globalMu.constructors
+	oldOptions := globalMu.options
+
+	globalMu.constructors = make(map[jobspb.Type]Constructor)
+	globalMu.options = make(map[jobspb.Type]registerOptions)
+	return func() {
+		globalMu.Lock()
+		defer globalMu.Unlock()
+		globalMu.constructors = oldConstructors
+		globalMu.options = oldOptions
+	}
+
+}
+
 // RegisterConstructor registers a Resumer constructor for a certain job type.
 //
 // NOTE: You must pass either jobs.UsesTenantCostControl or

--- a/pkg/jobs/registry_test.go
+++ b/pkg/jobs/registry_test.go
@@ -591,6 +591,9 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 		cancel = false
 	)
 
+	restoreConstructors := TestingClearConstructors()
+	defer restoreConstructors()
+
 	// createJob creates a mock job.
 	createJob := func(
 		t *testing.T, ctx context.Context, s serverutils.ApplicationLayerInterface, r *Registry, tdb *sqlutils.SQLRunner, db isql.DB,

--- a/pkg/jobs/registry_test.go
+++ b/pkg/jobs/registry_test.go
@@ -667,6 +667,7 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 		// expectImmediateRetry is true if the test should expect immediate
 		// resumption on retry, such as after pausing and resuming job.
 		expectImmediateRetry bool
+		allowCompletion      func()
 	}
 	testInfraSetUp := func(t *testing.T, ctx context.Context, bti *BackoffTestInfra) func() {
 		// We use a manual clock to control and evaluate job execution times.
@@ -785,15 +786,17 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 			require.GreaterOrEqual(t, maxDelay, delay, "delay exceeds the max")
 			// Advance the clock such that it is before the next expected retry time.
 			bti.clock.AdvanceTo(lastRun.Add(delay - unitTime))
-			// This allows adopt-loops to run for a few times, which ensures that
-			// adopt-loops do not resume jobs without correctly following the job
-			// schedules.
-			waitUntilCount(t, bti.adopted, bti.adopted.Count()+2)
 			if bti.expectImmediateRetry && i > 0 {
-				// Validate that the job did not wait to resume on retry.
+				// Validate that the job does not wait to resume on retry.
+				waitUntilCount(t, bti.resumed, expectedResumed+1)
 				require.Equal(t, expectedResumed+1, bti.resumed.Count(), "unexpected number of jobs resumed in retry %d", i)
 			} else {
 				// Validate that the job is not resumed yet.
+
+				// This allows adopt-loops to run for a few times, which ensures that
+				// adopt-loops do not resume jobs without correctly following the job
+				// schedules.
+				waitUntilCount(t, bti.adopted, bti.adopted.Count()+2)
 				require.Equal(t, expectedResumed, bti.resumed.Count(), "unexpected number of jobs resumed in retry %d", i)
 			}
 			// Advance the clock by delta from the expected time of next retry.
@@ -813,9 +816,13 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 			}
 			lastRun = bti.clock.Now()
 		}
-		bti.done.Store(true)
-		// Let the job be retried one more time.
-		bti.clock.Advance(nextDelay(retryCnt, initialDelay, maxDelay))
+
+		if bti.allowCompletion != nil {
+			bti.allowCompletion()
+		} else {
+			bti.done.Store(true)
+			bti.clock.Advance(nextDelay(retryCnt, initialDelay, maxDelay))
+		}
 		// Wait until the job completes.
 		testutils.SucceedsSoon(t, func() error {
 			var found Status
@@ -856,6 +863,11 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 	t.Run("pause running", func(t *testing.T) {
 		ctx := context.Background()
 		bti := BackoffTestInfra{expectImmediateRetry: true}
+		bti.allowCompletion = func() {
+			<-bti.resumeCh
+			bti.errCh <- nil
+			<-bti.transitionCh
+		}
 		bti.afterJobStateMachineKnob = func() {
 			if bti.done.Load().(bool) {
 				return
@@ -936,7 +948,11 @@ func TestRetriesWithExponentialBackoff(t *testing.T) {
 	t.Run("pause reverting", func(t *testing.T) {
 		ctx := context.Background()
 		bti := BackoffTestInfra{expectImmediateRetry: true}
-
+		bti.allowCompletion = func() {
+			<-bti.failOrCancelCh
+			bti.errCh <- nil
+			<-bti.transitionCh
+		}
 		bti.afterJobStateMachineKnob = func() {
 			if bti.done.Load().(bool) {
 				return

